src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractAsyncReceivedInOrderCallbackAnalyzer.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using NSubstitute.Analyzers.Shared.Extensions;
+
+namespace NSubstitute.Analyzers.Shared.DiagnosticAnalyzers
+{
+    internal abstract class AbstractAsyncReceivedInOrderCallbackAnalyzer<TSyntaxKind, TInvocationExpressionSyntax> : AbstractDiagnosticAnalyzer
```

Comment:
@dtchepak @zvirja this will detect async callback in following usages
```c#
Received.InOrder(async () => { // rest of the code })
Received.InOrder(async delegate { // rest of the code })
```
should I also make an attempt to detect async void callbacks like
```c#
Received.InOrder(Calls);
  
private async void Calls()
{
    await Task.CompletedTask;
}
```

Reply:
Maybe hold off for now. I haven't done much with async/await so I'm not sure, but maybe people will want to try asserting calls on a substitute that is returned from another async call:

```
private async void Calls() {
  var anotherSub = await sub.GetThingoe();
  anotherSub.SomeCall();
}
```

AFAICT the problem for NSubstitute is mainly when using `Received.InOrder(async () => ...)`; it seems ok without the `async` which I think is what we need to avoid. Keeping the option to extract this to a separate method might be a good workaround if people really need this. WDYT @zvirja ?

But I'm happy to be corrected on this as I only have basic experience with async/await.

Reply:
@dtchepak I like you idea of keeping `Received.InOrder(Calls);` for now in case clients have weird reason for that. This way we could also collect the feedback and see whether it's the case at all.

Also it should help keep code simpler and less chance of false-positive :)

Reply:
Thanks for your feedback guys. I am leaving this as it is